src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala

```diff
@@ -183,28 +180,27 @@ private class TaskLauncherActor(
 
   private[this] def receiveTaskLaunchNotification: Receive = {
     case InstanceOpSourceDelegate.InstanceOpRejected(op, TaskLauncherActor.OfferOperationRejectedTimeoutReason) =>
-      // Reschedule instance with provision timeout..
-      if (inFlightInstanceOperations.exists(_.instanceId == op.instanceId)) {
-        instanceMap.get(op.instanceId).foreach { instance =>
-          import scala.concurrent.ExecutionContext.Implicits.global
-
-          logger.info(s"Reschedule ${instance.instanceId} because of provision timeout.")
-          async {
-            if (instance.runSpec.isResident) {
-              await(instanceTracker.process(RescheduleReserved(instance.instanceId, instance.runSpec)))
-            } else {
-              // Forget about old instance and schedule new one.
-              await(instanceTracker.forceExpunge(instance.instanceId)): @silent
-              await(instanceTracker.schedule(Instance.scheduled(instance.runSpec)))
-            }
-          }.failed.map(Status.Failure).pipeTo(self)
-        }
+      if (inFlightInstanceOperations.exists(_.instanceId == op.instanceId) && offerOperationAcceptTimeout.contains(op.instanceId)) {
+        // TaskLauncherActor is not always in sync with the state in InstanceTracker.
+        // Especially when accepting offer, its state is ahead of InstanceTracker because it marks the instance as provisioned even before we persist it.
+        // We do this because we need to know that we already attempted to launch that instance to not use the same one for another offer.
+        // If everything goes well, some time after we manually adjust the internal state we should receive instance update to Provisioned.
+        // It might happen that the we were not able to persist that Provisioned operation for some reason.
+        // This timeout is here just to prevent us from staying in a state where we have different state locally than in InstanceTracker
+        // This timeout should be set that far in the future that the offer operation should have already timed out anyway so it should be safe to do.
+        // By explicitly syncing here we would either get the instance in Provisioned (we persisted the update) or scheduled (something failed).
+        syncInstance(op.instanceId)
```

Comment:
Nice! When I read your PR I thought that we just need to sync an instance. I think we might not even need the timeout but your solution is on the safe side.

Reply:
I agree it's nice, but I think it contains a bug :)

when we hit this, we should not only sync (i.e. most likely go back to a Scheduled instance), but also `manageOfferMatcherStatus(clock.now())` so that the actor registers as an offer matcher, if needed.

Hint: look at how these cases were handled before the refactoring. You should either add the `manageOfferMatcherStatus` to your first case, or do something similar to the original code:

**case 1:** we receive a provisioning timeout but the instance is not in flight anymore: **ignore the timeout**
_notice the `if` is part of the case statement!_
```
case InstanceOpRejected(op, OfferOperationRejectedTimeout) if (!inFlight(instanceId) && !offerOperationAcceptTimeout.contains(op.instanceId)) =>
  logger.debug("ignoring OfferOperationRejectedTimeout because the instance is not in flight anymore")
```

**case 2:** we receive a Rejected with any other reason, or for an instance that **is** in flight: **sync and check offer matcher status**
```
case InstanceOpSourceDelegate.InstanceOpRejected(op, reason) =>
  logger.debug(s"Task op '${op.getClass.getSimpleName}' for ${op.instanceId} was REJECTED, reason '$reason', rescheduling. $status")
  syncInstance(op.instanceId)
  manageOfferMatcherStatus(clock.now())
```

**case 3**: Accepted: **as implemented**
```
case InstanceOpSourceDelegate.InstanceOpAccepted(op) =>
  ...
```